README.md
[Simple Photos Contest](https://github.com/Dric/simple-photos-contest)
==========

A Simple photos contest gallery. Vote for your favorites photos.

50thomatoes50's fork ![Version-3.0 Alpha 5](https://img.shields.io/badge/Version-3.0%20Alpha%205-brightgreen.svg "Version-3.0 Alpha 5")
![MIT License](https://img.shields.io/github/license/50thomatoes50/simple-photos-contest.svg "MIT License")

Screenshots
-----------
![Frontend 1](https://raw.github.com/50thomatoes50/simple-photos-contest/master/install/img/front1.jpg "Frontend 1")
Look at the small cog icon in upper right corner : this is the settings button.

![Frontend 2](https://raw.github.com/50thomatoes50/simple-photos-contest/master/install/img/front2.jpg "Frontend 2")
Photos are automatically tiled. This can be a bit messy.

![Admin 1](https://raw.github.com/50thomatoes50/simple-photos-contest/master/install/img/admin1.jpg "Admin 1")
Admin panel - Contests list.

![Admin 2](https://raw.github.com/50thomatoes50/simple-photos-contest/master/install/img/admin2.jpg "Admin 2")
Admin panel - SPC settings bar.

Requirements
------------

+ PHP 5.3
+ MySQL 5.5
+ Gettext enabled (for translations)
+ GD Image Library

There is nothing unusual, as GD and gettext are enabled on most of web hosting platforms. Simple Photos Contest may work with previous versions of MySQL, but it hasn't been tested.

Runs well on a decent Internet Browser that support HTML5. Javascript is required.

Installation
------------
* Automatic install (recommended)
	1. Make the cache dir writable (777).
	2. Open the SPC url in your favorite browser.
	3. You will be redirected to the SPC installer.
	4. Follow the installer steps.
* Manual install :
	1. Make the cache dir writable (777).
	2. import install/install.sql in your mysql db.
	3. Rename config-sample.php into config.php.
	4. Edit config.php to change db connect values and admin password.
	5. Configure SPC in admin panel.


Update from 1.x versions
------------------------
Run SQL commands :

    ALTER TABLE `contests` ADD `voting_type` VARCHAR( 10 ) NOT NULL DEFAULT "open" ;
    ALTER TABLE `image_ip` CHANGE `ip_add` `ip_add` INT NULL DEFAULT NULL;

Update to 3.x  alpha versions
------------------------

### New method

- Always make a backup (files and SQL Database)
- Delete all files and directory that is not named "photos"
- Put the new file in
- Go to %yoursite%/install/upgrade.php with a browser
- Click 'Upgrade' and done

### Old method (not safe at all)

Run those SQL commands :

		ALTER TABLE `contests` ADD `icon` VARCHAR(200) NOT NULL ;
		ALTER TABLE `settings` ADD `language_auto` BOOLEAN NOT NULL ;
		ALTER TABLE `settings` ADD `homepage` BOOLEAN NOT NULL ;
		ALTER TABLE `settings` ADD `auth_method` INT(2) NOT NULL , ADD `spc_version` VARCHAR(8) NOT NULL ;

Quick Start
-----------

+ Create directory in your photos dir.
+ Put your photos into this dir.
+ Activate it in admin panel.
+ The first load of a contest page is slow. No panic, it's because SPC needs to create thumbnails of all the gallery photos. The next load will be a lot faster !

FAQ
---

<dl>
	<dt>Where is the admin panel ?</dt>
	<dd>There is small icon in upper right corner of the main page. Just click on it, enter password and you will get the admin panel.</dd>
	<dt>How can I translate Simple Photos contest ?</dt>
	<dd>There is a pot template file in lang dir. Edit it with a po editor like [PoEdit](http://www.poedit.net) in Windows, then mimic the french translation files tree.</dd>
	<dt>How can I change SPC CSS ?</dt>
  <dd>You should modify SPC LESS files (check `variables.less` first, as most of colors and sizes are defined here) and compile it to a CSS file.</dd>
</dl>

TO-DO
-----

+ Use of responsive template
+ Use of HTML5 + CSS3 to replace jQuery animations

Components used
---------------

+ [SimpleImage](https://github.com/claviska/SimpleImage) 2.4
+ [jQuery](http://jquery.com) 1.11.2
+ [Freetile](https://github.com/yconst/Freetile) 0.3.1
+ [Lightbox](http://lokeshdhakar.com/projects/lightbox2) 2.8.1
+ [glDatePicker](http://glad.github.io/glDatePicker) 2.0
+ [jqBarGraph](http://workshop.rs/jqbargraph/) 1.1
+ [Tiny Scrollbar](http://baijs.nl/tinyscrollbar/) 1.81
+ [less.php](http://lessphp.gpeasy.com/) 1.7.0.5
+ [Font Awesome](http://fortawesome.github.io/Font-Awesome/) 4.4.0
+ [SteamAuthentication](https://github.com/SmItH197/SteamAuthentication)


Bug tracker
-----------

Have a bug? Being shocked by a poor translation (I'm french) ? Please create an issue here on GitHub !

<https://github.com/Dric/simple-photos-contest/issues>


Author
-------

**Dric**

+ Blog : <http://www.driczone.net>
+ Twitter : <http://twitter.com/Dric>
+ Github : <http://github.com/Dric>


Copyright and license
---------------------

Simple Photos Contest is released under MIT licence.

The MIT License (MIT)

Copyright (c) 2012-2015 Dric

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"),
to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software,
and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.MD
Simple Photos Contest
==========

ChangeLog
--------

* 3.0 Alpha 5
  - Enhancement of the user page
    - Table padding Added
    - Image preview for votes (gallery and voted)


* 3.0 Alpha 4
  - Fixed some install bug
    - Used an incorrect path for config-sample.php
    - Added a message when the config.php file is generated but empty
    - Forgot a ";" in install.sql
  - Fixed a bug when reseting votes
  - Fixed Vote type in admin not selecting the correct voting type
  - Fixed user page always showing "didn't vote"


* 3.0 Alpha 3
  - Updated lightbox to latest version
  - Added missing images file for lightbox
  - Fixed "Lightbox is bugged #8"


* 3.0 Alpha 2
  - Theme (Change color of variable.less and recompile it in the admin panel)
  - Move included file to a secure folder
  - updated AwesomeFont to 4.4
  - LogOut bouton for admin page
  - SteamAuthentication added
  - fixed an exploit (can vote whenever a contest is expired)
  - Installer can upgrade the db when upgrading SPC
  - Automatic language selector (not tested)


* 3.0 Alpha 1
  - You can add icon to a gallery
  - Added a proper main page
  - Added some button to go back to main page


* 2.0
	- Major rewrite : frontend now use HTML5 and CSS3
	- Used of LESS files to build CSS file
	- Use fingerprinting method instead of IP address to detect if a visitor has already voted.
	- ZebraDatePicker replaced by glDatePicker


* 1.3
	- `mysql_*` obsolete php functions replaced by `mysqli_*` functions.
	- When `config.php` file is not found, user will be redirected to install script.
	- Link to admin panel in breadcrumb
	- When contests are saved in Db but no default contest is set, SPC will display the first contest saved in Db
	- Better SQL error check
	- A few bugs fixed
	- Updated jQuery to v1.11.2
	- Updated freeTile to v0.3.1
	- Custom Slimbox2 replaced by LightBox


* 1.2
  - Use of SimpleImage class instead of TimThumb
  - Add a setting to allow voting once in a contest or unlimited votes (you are still not allowed to vote more than once for a photo)


functions.php
<?php 
DEFINE('SPC_VERSION', '3.0 Alpha 1');

$sql=mysqli_query($bd, "SELECT * FROM settings");
$settings = mysqli_fetch_object($sql);
unset($sql);

if (!empty($settings)){
	/** Translations ! */

	/** Define language used
	*
	* To see the locales installed in your ubuntu server, type locale -a in shell.
	*/
	putenv("LC_ALL=".$settings->language);
	setlocale(LC_ALL, $settings->language);
	bindtextdomain("messages", "lang");
	bind_textdomain_codeset('messages', 'UTF-8');
	textdomain("messages");

}

/** Are we logged in ? */
$admin_logged = admin_logged();


/**
* Transcrit une date au format sql en format français.
*
* @param string $mysql_date DATE ou DATETIME SQL
* @param string $time Si valeur égale à 'notime' (valeur par défaut), on ne retourne pas les heures:minutes:secondes.
*/
function date_formatting($mysql_date, $to_sql = false){
	global $settings;
  if (!$to_sql){
    //return date_format(date_create($mysql_date), $settings->date_format);
		return changeDateFormat($mysql_date, 'Y-m-d', $settings->date_format);
  }else{
    //return date_format(date_create_from_format($settings->date_format, $mysql_date), 'Y-m-d');
		return changeDateFormat($mysql_date, $settings->date_format, 'Y-m-d');
  }
}
/** For php version < 5.3 
* http://php.net/manual/en/function.date.php#90423
*/
function dateParseFromFormat($stFormat, $stData)
 {
     $aDataRet = array('day'=>0, 'month'=>0, 'year'=>0, 'hour'=>0, 'minute'=>0, 'second'=>0);
     $aPieces = preg_split('[:/.\ \-]', $stFormat);
     $aDatePart = preg_split('[:/.\ \-]', $stData);
     foreach($aPieces as $key=>$chPiece)    
     {
         switch ($chPiece)
         {
             case 'd':
             case 'j':
                 $aDataRet['day'] = $aDatePart[$key];
                 break;
                 
             case 'F':
             case 'M':
             case 'm':
             case 'n':
                 $aDataRet['month'] = $aDatePart[$key];
                 break;
                 
             case 'o':
             case 'Y':
             case 'y':
                 $aDataRet['year'] = $aDatePart[$key];
                 break;
             
             case 'g':
             case 'G':
             case 'h':
             case 'H':
                 $aDataRet['hour'] = $aDatePart[$key];
                 break;    
                 
             case 'i':
                 $aDataRet['minute'] = $aDatePart[$key];
                 break;
                 
             case 's':
                 $aDataRet['second'] = $aDatePart[$key];
                 break;            
         }
         
     }
     return $aDataRet;
 }
 
 function changeDateFormat($stDate,$stFormatFrom,$stFormatTo)
 {
   // When PHP 5.3.0 becomes available to me
   $date = date_parse_from_format($stFormatFrom,$stDate);
   //For now I use the function above
   //$date = dateParseFromFormat($stFormatFrom,$stDate);
   return date($stFormatTo,mktime($date['hour'],
                                     $date['minute'],
                                     $date['second'],
                                     $date['month'],
                                     $date['day'],
                                     $date['year']));
 }

/**
* Valide une date suivant un format.
* 
* @param string $date Date à valider
* @param string $format Format que doit avoir la date à valider
*/
function date_valid($date, $format = null) {
	global $settings;
	if (empty($format)){
		$format = $settings->date_format;
	}
   if (date($format, strtotime($date)) == $date) {
       return true;
   } else {
       return false;
   }
 }
 
/**
 * Permet de savoir si l'admin est connecté.
 *
 * @return bool
 */
function admin_logged(){
  if (isset($_COOKIE[COOKIE_NAME])){
    $cookie = $_COOKIE[COOKIE_NAME];
    if ($cookie == sha1(PASSWD.HASH)){
			return true;
		}
  }
	return false;
}
/**
 * Affiche un message de notification.
 *
 * @param object $message
 */
function disp_message($message){
	if (isset($message->type)){
		return '<div class="alert '.$message->type.'"><a class="alert-close" href="#" title="Fermer">×</a>'.$message->text.'</div>';
	}
}

function info_disp($message){
	return '<span class="fa fa-info imgInfo" title="'.$message.'"></span>';
	//<img alt="'.$message.'" class="img_info" src="img/info.png" />';
}

/**
* Display a vertical bar graph with contest votes.
* @param string $contest Contest ID
* 
*/
function contest_stats($contest){
	global $settings, $c_path, $bd;
	// Some text to explain why the voting system is not 100% reliable
	?><div class="width60Percent"><?php echo _('Warning : SPC does not provide voters authentication, and as a result the voting limit is not 100% reliable. People could vote multiple times with other devices or browsers and SPC would not be able to detect that this is the same voter.'); ?></div><?php
	/** Get contest data. */
	$sql=mysqli_query($bd, 'SELECT * FROM contests WHERE contest = "'.$contest.'"');
	$cont = mysqli_fetch_object($sql);
	/** Get images and votes. */
	$sql=mysqli_query($bd, 'SELECT * FROM images WHERE contest = "'.$contest.'" ORDER BY img_name');
	$nbphotos = mysqli_num_rows($sql);
	/** Get number of voters (format : array with only first value populated) */
	$nbvoters = mysqli_fetch_row(mysqli_query($bd, 'SELECT COUNT(DISTINCT ip_add) FROM image_IP WHERE contest = "'.$contest.'"'));
	$nbvoters = $nbvoters[0];
	/** Let's build the graph source js array ! */
	?>
	<script>
		if (typeof arrayOfData == "undefined") {
			arrayOfData = new Array();
		}
		arrayOfData["<?php echo $contest; ?>"] = new Array(
	<?php
	$disp = '';
	$nbvotes = 0;
	/** @param array Photo name as first value, img url as second and number of votes as third. */
	$mostvoted = array(0,0,0);
	while($row=mysqli_fetch_array($sql)){
		$disp .= "\r\n".'['.$row['love'].', "<a title=\"'.$row['img_name'].'\" href=\"'.$c_path.$contest.'/'.$row['img_url'].'\" class=\"\" data-lightbox=\"'.$row['img_id'].'\">'.$row['img_name'].'</a>"],';
		$nbvotes += $row['love'];
		if ($mostvoted[2] < $row['love']){
			$mostvoted[0] = $row['img_name'];
			$mostvoted[1] = $row['img_url'];
			$mostvoted[2] = $row['love'];
		}
	}
	$disp = trim($disp, ',');
	echo $disp;
	?>
		);
	</script>
	<div class="graph" data-contest="<?php echo $contest; ?>" data-title="<h2><?php echo sprintf(_('%s contest'), $cont->contest_name); ?></h2>" id="contest_graph_<?php echo $contest; ?>"></div>
	<ul class="stats_data">
		<li><?php echo _('Number of votes'); ?> : <span class="stats_numbers"><?php echo $nbvotes; ?></span></li>
		<li><?php echo _('Number of voters'); ?> : <span class="stats_numbers"><?php echo $nbvoters; ?></span></li>
		<li><?php echo _('Number of photos'); ?> : <span class="stats_numbers"><?php echo $nbphotos; ?></span></li>
		<li><?php echo _('Favorite'); ?> : <span class="stats_numbers"><a class="" data-lightbox="favorite" href="<?php echo $c_path.$contest.'/'.$mostvoted[1]; ?>" title="<?php echo $mostvoted[0]; ?>"><?php echo $mostvoted[0]; ?></a></span> <?php echo _('with'); ?> <span class="stats_numbers"><?php echo $mostvoted[2]; ?></span> <?php echo ngettext('vote', 'votes', $mostvoted[2]); ?></li>
	</ul>
	<?php
}
?>